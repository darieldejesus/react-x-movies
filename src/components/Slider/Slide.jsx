import React, { Component } from 'react';
import Styled, { css } from 'styled-components';
import PropTypes from 'prop-types';

const Container = Styled.div`
  background: #FFF;
  display: flex;
  float: left;
  margin-right: -100%;
  opacity: 0;
  transition: opacity 1s;
  width: 100%;
  z-index: 1;

  ${props => props.active && css`
    opacity: 1;
    z-index: 2;
  `}
`;

const Image = Styled.img`
  flex: 1;
`;

const DetailsContainer = Styled.div`
  flex: 1;
  padding: 10px;
`;

const Header = Styled.h1`
  color: #ed0105;
  font-size: 2em;
`;

const Details = ({
  title,
  date,
  rate,
  overview,
}) => (
  <DetailsContainer>
    <Header>{ title }</Header>
    <p><b>Release: </b>{date}</p>
    <p><b>Rating: </b>{rate}</p>
    <p>{overview}</p>
  </DetailsContainer>
);

Details.propTypes = {
  date: PropTypes.string.isRequired,
  overview: PropTypes.string.isRequired,
  rate: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
};

class Slide extends Component {
  constructor() {
    super();

    this.state = {
      loaded: false,
    };
  }

  componentDidMount() {
    if (this.props.active) {
      this.setState({ loaded: true });
    }
  }

  componentWillReceiveProps() {
    if (!this.state.loaded && this.props.active) {
      this.setState({ loaded: true });
    }
  }

  render() {
    const {
      active, title, image, preLoad, overview, release, rate,
    } = this.props;
    const { loaded } = this.state;

    return (
      <Container active={active}>
        {(loaded || active || preLoad)
        && [
          <Image key="slide-image" src={`https://image.tmdb.org/t/p/w780/${image}`} alt={title} />,
          <Details key="slide-details" title={title} date={release} rate={rate} overview={overview} />]
        }
      </Container>
    );
  }
}

Slide.propTypes = {
  active: PropTypes.bool.isRequired,
  image: PropTypes.string.isRequired,
  overview: PropTypes.string.isRequired,
  preLoad: PropTypes.bool.isRequired,
  rate: PropTypes.number.isRequired,
  release: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
};

export default Slide;
