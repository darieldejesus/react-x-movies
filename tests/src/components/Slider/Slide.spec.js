import React from 'react';
import Slide from '../../../../src/components/Slider/Slide';
import fixture from '../../../fixture/popularMovies.json';

describe('Slide', function () {
  beforeEach(() => {
    const entry = fixture[0];
    this.wrapper = shallow(<Slide
      active={true}
      key={entry.id}
      image={entry.backdrop_path}
      overview={entry.overview}
      preLoad={true}
      rate={entry.vote_average}
      release={entry.release_date}
      title={entry.title}
    />);
  });

  test('should renders Slide', () => {
    expect(toJson(this.wrapper)).toMatchSnapshot();
  });
});
