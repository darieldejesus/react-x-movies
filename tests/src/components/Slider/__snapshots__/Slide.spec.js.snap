// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide should renders Slide 1`] = `
<styled.div
  active={true}
>
  <styled.img
    alt="It"
    key="slide-image"
    src="https://image.tmdb.org/t/p/w780//tcheoA2nPATCm2vvXw2hVQoaEFD.jpg"
  />
  <Details
    date="2017-09-05"
    key="slide-details"
    overview="In a small town in Maine, seven children known as The Losers Club come face to face with life problems, bullies and a monster that takes the shape of a clown called Pennywise."
    rate={7.3}
    title="It"
  />
</styled.div>
`;
